### PR TITLE
Add VitMatte node

### PR DESCRIPTION
**Pull request by demoulinv**
This PR adds a new matting node named ViTMatte and adjust the computation of the binary mask computed from the full box (without tiling) in the VideoSegmentationSam3Boxes node.

### CODE REVIEW

File: meshroom/imageSegmentation/ViTMatte.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
+__version__ = "1.0"
+
+from chunk import Chunk
+import chunk
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]:
```suggestion

```

File: meshroom/imageSegmentation/ViTMatte.py [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+import logging
+logger = logging.getLogger("ViTMatte")
+
+class ViTMatte(desc.Node):
+    size = avpar.DynamicViewsSize("input")
+    gpu = desc.Level.INTENSIVE
+    parallelization = desc.Parallelization(blockSize=50)
+
+    category = "Matting"
+    documentation = """
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py:19]: The `documentation` field is not necessary anymore, we can directly use the class' docstring for the documentation now.
```suggestion
```

File: meshroom/imageSegmentation/ViTMatte.py [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+            label="Trimap Computation Method",
+            description="Method to compute a trimap from a binary mask.",
+            value="Blur",
+            values=["Blur", "Morpho"],
+            exclusive=True,
+        ),
+        desc.IntParam(
+            name="kernelSize",
+            label="Kernel Size",
+            description="Gaussian blur kernel size or Erode/Dilate rectangle kernel size",
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]:
```suggestion
            description="Gaussian blur kernel size or Erode/Dilate rectangle kernel size.",
```

File: meshroom/imageSegmentation/ViTMatte.py [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+            pathMask = ""
+            promptType = ""
+            if chunk.node.inputMask.value != "":
+                pathMask = chunk.node.inputMask.value
+                promptType = "mask"
+            elif chunk.node.inputTrimap.value != "":
+                pathMask = chunk.node.inputTrimap.value
+                promptType = "trimap"
+
+            modelInfo= {"Comp1k": {True: "ViTMatte_B_Com.pth", False: "ViTMatte_S_Com.pth"},
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]:
```suggestion
            modelInfo = {"Comp1k": {True: "ViTMatte_B_Com.pth", False: "ViTMatte_S_Com.pth"},
```

File: meshroom/imageSegmentation/ViTMatte.py [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+
+            outFiles = self.resolvedPaths(chunk.node.input.value,
+                                          pathMask, chunk.node.extensionMask.value,
+                                          chunk.node.output.value, chunk.node.keepFilename.value,
+                                          chunk.node.extensionOut.value)
+
+            if promptType == "":
+                raise ValueError("Some images have no valid mask or trimap to drive the matting process !!!")
+            else:
+                logger.info(f"prompt type: {promptType}")
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]: This would be better if placed before the call to `self.resolvedPaths`, since there is no point in calling it if an exception is going to be raised right after.

```suggestion
            if promptType == "":
                raise ValueError("Some images have no valid mask or trimap to drive the matting process!")
            else:
                logger.info(f"Prompt type: {promptType}")
```

File: meshroom/imageSegmentation/ViTMatte.py [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+
+                if not os.path.exists(chunk.node.output.value):
+                    os.mkdir(chunk.node.output.value)
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]: Real question: do we ever enter in that `if`?

Since we're in the `processChunk`, we should already have at least the log/status files in the cache folder, there should not be a case where it hasn't been created yet.

File: meshroom/imageSegmentation/ViTMatte.py [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+        if Path(pathIn).suffix.lower() in [".sfm", ".abc"]:
+            if Path(pathIn).exists():
+                dataAV = sfmData.SfMData()
+                if sfmDataIO.load(dataAV, pathIn, sfmDataIO.ALL) and os.path.isdir(outDir):
+                    views = dataAV.getViews()
+                    for id, v in views.items():
+                        inputFile = v.getImage().getImagePath()
+                        frameId = v.getFrameId()
+                        if keepFilename:
+                            if pathMask != "":
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]:
```suggestion
                            if pathMask:
```

File: meshroom/imageSegmentation/ViTMatte.py [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+                                inputFileMask = os.path.join(pathMask, Path(inputFile).stem + "." + extMask)
+                            outputFileMatte = os.path.join(outDir, Path(inputFile).stem + "." + extOut)
+                            outputFileTrimap = os.path.join(outDir, "trimap_" + Path(inputFile).stem + "." + extOut)
+                        else:
+                            if pathMask != "":
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]:
```suggestion
                            if pathMask:
```

File: meshroom/imageSegmentation/ViTMatte.py [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+        import OpenImageIO as oiio
+
+        try:
+            logger.setLevel(chunk.node.verboseLevel.value.upper())
+
+            if not chunk.node.input:
+                logger.warning("Nothing to segment")
+                return
+            if not chunk.node.output.value:
+                return
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]:
```suggestion
```
Useless since `output` has a set value in the description, and is thus read-only (so there is no chance of it ever being empty).

File: meshroom/imageSegmentation/ViTMatte.py [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+        import numpy as np
+        import torch
+        from pyalicevision import image as avimg
+        import OpenImageIO as oiio
+
+        try:
+            logger.setLevel(chunk.node.verboseLevel.value.upper())
+
+            if not chunk.node.input:
+                logger.warning("Nothing to segment")
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]: Wouldn't it be better to raise here? If we don't, then the node does not process anything but will still end up with a `SUCCESS` status, which I'm not sure is what we should aim for.

Raising should at least update it to the `ERROR` status.

```suggestion
                raise ValueError("Nothing to segment.")
```

File: meshroom/imageSegmentation/ViTMatte.py [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+                                return
+
+                            ys = coords[:, 0]
+                            xs = coords[:, 1]
+                            x_left = xs.min()
+                            x_right = xs.max() + 1
+                            y_top = ys.min()
+                            y_bottom = ys.max() + 1
+                            x_center = int((x_right + x_left) / 2)
+                            y_center = int((y_top + y_bottom) / 2)
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]: Wouldn't it be faster to do `(x_right + x_left) // 2` and `(y_top + y_bottom) // 2` rather than doing a float division and then casting it back to int?

File: meshroom/imageSegmentation/ViTMatte.py [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+                            width = x_right - x_left
+                            height = y_bottom - y_top
+
+                            trimap_ext = 1.1
+                            mask_height, mask_width, _ = promptRGB.shape
+                            x_left_inference = max(0, x_center - int(trimap_ext * width / 2.0))
+                            x_right_inference = min(mask_width, x_center + int(trimap_ext * width / 2.0))
+                            y_top_inference = max(0, y_center - int(trimap_ext * height / 2.0))
+                            y_bottom_inference = min(mask_height, y_center + int(trimap_ext * height / 2.0))
+                            top_left_xy     = (x_left_inference, y_top_inference)
```
Comment (cbentejac) [inline meshroom/imageSegmentation/ViTMatte.py]:
```suggestion
                            top_left_xy = (x_left_inference, y_top_inference)
```

**Review by cbentejac — CHANGES_REQUESTED** [review]

**Review by cbentejac — APPROVED** [review]
